### ISSUE TITLE
LSP completion: display documentation if exists for an item

### ISSUE THREAD

**Issue by przepompownia** [neovim/neovim#29225]
### Problem

It seems that at the moment displaying documentation for LSP completion does not work. If I'm wrong please close this issue. 

We probably need to call `[completionItem/resolve](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem_resolve)` for each item. The only such call seems to be made for single submitted item. 

### Expected behavior

Display documentation attached to an item - maybe by the user choice with some new option on `vim.lsp.completion.BufferOpts`.

I have created a simple [example](https://gist.github.com/przepompownia/0690ebe28a24bd10b45118fceb980dfd) containing autocommand on `CompleteChanged` that shows that it's not difficult but not sure how it should like as a built-in functionality so I didn't create PR. 
 
```lua
vim.api.nvim_create_autocmd('FileType', {
  pattern = 'lua',
  callback = function (args)
    vim.treesitter.start(args.buf, 'lua')
    vim.lsp.start({
      name = 'lua-ls',
      cmd = {'lua-language-server'},
      root_dir = vim.fs.root(args.buf, 'init.lua'),
    })
  end,
})

vim.api.nvim_create_autocmd('LspAttach', {
  callback = function (args)
    local client = vim.lsp.get_client_by_id(args.data.client_id)

    vim.lsp.completion.enable(true, args.data.client_id, args.buf, {autotrigger = true})

    vim.keymap.set({'i'}, '<C-Space>', function ()
      vim.lsp.completion.trigger()
    end, {buffer = args.buf})

    vim.api.nvim_create_autocmd('CompleteChanged', {
      buffer = args.buf,
      callback = function ()
        local info = vim.fn.complete_info({'selected'})
        local completionItem = vim.tbl_get(vim.v.completed_item, 'user_data', 'nvim', 'lsp', 'completion_item')
        if nil == completionItem then
          return
        end

        client.request(vim.lsp.protocol.Methods.completionItem_resolve, completionItem, function (_err, result)
          if nil == result then
            return
          end

          local winData = vim.api.nvim__complete_set(info['selected'], {info = result.documentation.value})
          if not vim.api.nvim_win_is_valid(winData.winid) then
            return
          end
          vim.api.nvim_win_set_config(winData.winid, {border = 'rounded'})
          vim.treesitter.start(winData.bufnr, 'markdown')
          vim.wo[winData.winid].conceallevel = 3
        end, args.buf)

        vim.api.nvim_create_autocmd({'TextChangedI'}, {
          buffer = args.buf,
          callback = function ()
            vim.lsp.completion.trigger()
          end
        })
      end
    })
  end
})

vim.go.completeopt = 'menu,noselect,menuone,popup'

local pumMaps = {
  ['<Tab>'] = '<C-n>',
  ['<Down>'] = '<C-n>',
  ['<S-Tab>'] = '<C-p>',
  ['<Up>'] = '<C-p>',
  ['<CR>'] = '<C-y>',
}

for insertKmap, pumKmap in pairs(pumMaps) do
  vim.keymap.set(
    {'i'},
    insertKmap,
    function ()
      return vim.fn.pumvisible() == 1 and pumKmap or insertKmap
    end,
    {expr = true}
  )
end
```

**Comment by glepnir** [comment]
request is asynchronous, it is best to use a timer to ensure that selected is the same before you invoke nvim__set_complete. and when no result and you also need close exist info window.

**Comment by przepompownia** [comment]
I then have updated the example to
```lua
vim.api.nvim_create_autocmd('FileType', {
  pattern = 'lua',
  callback = function (args)
    vim.treesitter.start(args.buf, 'lua')
    vim.lsp.start({
      name = 'lua-ls',
      cmd = {'lua-language-server'},
      root_dir = vim.fs.root(args.buf, 'init.lua'),
    })
  end,
})

vim.api.nvim_create_autocmd('LspAttach', {
  callback = function (args)
    vim.lsp.completion.enable(true, args.data.client_id, args.buf, {autotrigger = true})

    vim.keymap.set({'i'}, '<C-Space>', function ()
      vim.lsp.completion.trigger()
    end, {buffer = args.buf})

    vim.api.nvim_create_autocmd('CompleteChanged', {
      buffer = args.buf,
      callback = function ()
        local info = vim.fn.complete_info({'selected'})
        local completionItem = vim.tbl_get(vim.v.completed_item, 'user_data', 'nvim', 'lsp', 'completion_item')
        if nil == completionItem then
          return
        end

        local resolvedItem = vim.lsp.buf_request_sync(
          args.buf,
          vim.lsp.protocol.Methods.completionItem_resolve,
          completionItem,
          500
        )

        local docs = vim.tbl_get(resolvedItem[args.data.client_id], 'result', 'documentation', 'value')
        if nil == docs then
          return
        end

        local winData = vim.api.nvim__complete_set(info['selected'], {info = docs})
        if not winData.winid or not vim.api.nvim_win_is_valid(winData.winid) then
          return
        end

        vim.api.nvim_win_set_config(winData.winid, {border = 'rounded'})
        vim.treesitter.start(winData.bufnr, 'markdown')
        vim.wo[winData.winid].conceallevel = 3

        vim.api.nvim_create_autocmd({'TextChangedI'}, {
          buffer = args.buf,
          callback = function ()
            vim.lsp.completion.trigger()
          end
        })
      end
    })
  end
})

vim.go.completeopt = 'menu,noselect,menuone,popup'

local pumMaps = {
  ['<Tab>'] = '<C-n>',
  ['<Down>'] = '<C-n>',
  ['<S-Tab>'] = '<C-p>',
  ['<Up>'] = '<C-p>',
  ['<CR>'] = '<C-y>',
}

for insertKmap, pumKmap in pairs(pumMaps) do
  vim.keymap.set(
    {'i'},
    insertKmap,
    function ()
      return vim.fn.pumvisible() == 1 and pumKmap or insertKmap
    end,
    {expr = true}
  )
end
```

**Comment by glepnir** [comment]
Not sure if it's required for the core. But you can definitely use it as a plugin.

**Comment by przepompownia** [comment]
> Not sure if it's required for the core. But you can definitely use it as a plugin.

Of course, but if so, I need an explanation why we would like to end with delivering `vim.lsp.completion` in its current form, which is useful, but only partially (and especially why its implementation landed here instead in some modern plugin). Have we discussed this somewhere?

**Comment by flexagoon** [comment]
I agree that this should be added in core. I think that directly sending LSP requests is not the level of abstraction that a user should touch to get basic functionality.

The current LSP completion API provides an abstraction on top of LSP requests which is usable out of the box, and if you need to change the completion items, or write custom trigger logic, you can do all of that with simple Lua code, without messing with underlying LSP requests.

However, the only way to get completion documentation is to manually envoke an LSP request, which is, in my opinion, too low level for something that is generally considered a basic part of the autocompletion experience.

I think the default LSP completion, while being basic, should still be fully usable out of the box, and it mostly is. Here's my configuration:

https://github.com/flexagoon/init.lua/blob/main/lua/config/completion.lua

Most of that code is just me defining custom trigger behavior since I don't like the default one, but the completion is perfectly usable even without that.

**Comment by crwebb85** [comment]
In my config, I have implemented `completionItem.resolveSupport` lsp client capability for the documentation field and markdown `completionItem.documentationFormat` client capability. I was going to create new issue but found this issue already exists. I would be willing to work on a PR to implement it in neovim. However, I want to get some feedback on my approach and discuss several bugs/missing features that resulted in a more complicated implementation. I will create issues where needed for those bugs/features requests and reference them to this issue in a day or so.

My intent with this comment is to:

1. Define specific feature requirements (to govern at least my proposed implementation)
2. Show it is at least possible to implement these requirements without any C code changes
3. Highlight pain-points by showing the lua workarounds
4. Propose potential solutions to those pain-points that could be solved with changes to the C code. When appropriate, I will create seperate issues for them and link back to this issue

Because there is a lot of name collisions in the documentation/API for these features,
I try to adhere to the following terminology:

- Preview window/buffer: the normal vim preview window/buffer that can be open with `:pedit`
- Preview documention window/buffer: the preview window/buffer when it displays completion documentation
- Floating documentation window/buffer: the floating window/buffer that displays the completion documentation (I have seen at least three different ways to refer to this but they all have name collisions with other things)
- Documentation window/buffer: either the preview documentation window/buffer or the floating documention window/buffer depending on what the user has their vim.bo.completeopt set to. I will use this when I need a collective term to refer to them.
- Documentation: the contents that is or could be (depending on context) displayed in the documentation buffer
- Documentation field: the documentation field of the completion item defined in the lsp spec
- Completion menu: the popup listing completion items

Requirements:

- Must respect `vim.bo.completopt` settings of `popup`, `preview`, or neither
  - If `popup`, use the floating documention window
  - If `preview`, use the preview documetion window
  - If both `popup` and `preview`, use the floating documention window to match current behavior
  - If neither `popup` or `preview`, do not display any documention window
- Must set documentation buffer filetype set to `markdown` if lsp documentation kind is markdown
- Must clear the documentation buffer filetype when lsp completion item documentation kind is not markdown
- Must display the correct tressiter highlights in the documentation buffer for the filetype
- Must fetch the documentation with a `completionItem/resolve` request and display the results in the documetion window when the lsp for the completion item supports `completionItem/resolve`, and the info field is missing from completion item.
- Must show the previously displayed documentation when the active completion changes to an item without documention (this is how it currently functions and my implementation should not change it despite not being my preference)
- Must follow project conventions (this is the one requirement my init.lua will ignore since it's purpose differs from an actual PR)

Pain Points:

- Getting the preview winid and preview bufnr is significantly more work when `completeopt` is set to `preview`
- Treesitter bugs result in hacks where I always close the preview/popup buffer and reopen it (there is an active PR that will fix these issues)
- Having to use the experimental api `nvim__complete_set` is concerning but likely could be refactored out whenever an official way of updating the displayed documentation is settled on.
- Needing to manually create a preview window to display the documentation in when `completeopt` is set to `preview` feels bad compared to the floating documention window which is displayed on a simple call to `nvim__complete_set`
- Not having a field in the vim completion item for the documentation filetype results in always having to checking the `user_data.nvim.lsp.completion_item.documentation` field. Between that and the my workarounds for the Treesitter issues it was just easier/simpler to display the documentation from the `user_data.nvim.lsp.completion_item.documentation` field rather then the `info` field which is an awkward [desire path](https://en.wikipedia.org/wiki/Desire_path).
- The help docs has a lot of name collisions for the completion documentation making looking up information and discussing it difficult.

## Implementation and Tests

Below is my implementation of those requirements in lua and a dummy lsp server to test it. The documentation window will say what feature I was testing for that displayed completion item.
Note: I did not go out of my way to test and prevent race conditions. That's a future problem for future me.

1. Copy the below code to an init.lua file
2. Make sure escaped backticks are unescaped (using the copy button on the codeblock should properly unescape the backticks)
3. Run `nvim --clean -u init.lua init.lua`
4. Go into insert mode with `i`
5. Type `test`
6. Open completion menu with `<c-x><c-o>`
7. Navigate up or down the list of with either the arrow keys or `<c-n>`
8. Retest with different completopt values by uncommenting the one you want to test

```lua
-------------------------------------------------------------------------------
---Setup options (you can swap between the three different versions)

-- Shows documentation floating window
vim.o.completeopt = "menu,menuone,noinsert,noselect,popup"

-- Shows documentation preview window
-- vim.o.completeopt = "menu,menuone,noinsert,noselect,preview"

-- Does not show documentation window
-- vim.o.completeopt = "menu,menuone,noinsert,noselect"

-------------------------------------------------------------------------------
--- My implementation of documentation completionItem/resolve

local function is_documentation_disabled()
	local options = vim.opt.completeopt:get()
	for _, option in ipairs(options) do
		if option == "popup" then
			return false
		end
		if option == "preview" then
			return false
		end
	end
	return true
end

local function get_preview_type()
	local type = nil
	local options = vim.opt.completeopt:get()
	for _, option in ipairs(options) do
		if option == "popup" then
			return "popup"
		end
		if option == "preview" then
			type = "preview"
		end
	end
	return type
end

---@class PreviewCompleteInfo
---@field selected integer
---@field preview_bufnr? integer
---@field preview_winid? integer

---@return PreviewCompleteInfo
local function get_preview_info()
	local complete_info = vim.fn.complete_info({
		"selected",
		"preview_winid",
		"preview_bufnr",
	})

	if get_preview_type() == "preview" then
		for _, winid in ipairs(vim.api.nvim_tabpage_list_wins(0)) do
			local wininfo = vim.fn.getwininfo(winid)[1]
			local window_type = vim.fn.win_gettype(wininfo.winnr)
			if window_type == "preview" then
				complete_info.preview_winid = winid
				complete_info.preview_bufnr = wininfo.bufnr
				return complete_info
			end
		end
	end
	return complete_info
end

---@param documentation string|lsp.MarkupContent|nil
local function is_lsp_completion_documention_empty(documentation)
	return documentation == nil
		or documentation == ""
		or (type(documentation) == table and documentation.value == "")
		or (type(documentation) == table and documentation.value == nil)
end

---@param lsp_documentation string|lsp.MarkupContent
---@param selected integer index of selected completion item 0-indexed
local function display_documentation_popup(lsp_documentation, selected)
	local kind = lsp_documentation and lsp_documentation.kind
	local filetype = kind == "markdown" and "markdown" or ""
	local documentation_value = type(lsp_documentation) == "string" and lsp_documentation or lsp_documentation.value
	local preview_info = vim.api.nvim__complete_set(selected, {
		info = documentation_value,
	})
	if preview_info.bufnr ~= nil then
		vim.bo[preview_info.bufnr].filetype = filetype
		if filetype == "markdown" then
			vim.treesitter.start(preview_info.bufnr)
		end
	end
end

---@param lsp_documentation string|lsp.MarkupContent
local function display_documentation_preview(lsp_documentation)
	local kind = lsp_documentation and lsp_documentation.kind
	local filetype = kind == "markdown" and "markdown" or ""
	local documentation_value = type(lsp_documentation) == "string" and lsp_documentation or lsp_documentation.value

	local preview_buf_name = "Documentation\\ Preview"
	vim.cmd(
		"silent! pedit! +setlocal\\ buftype=nofile\\ nobuflisted\\ noswapfile\\ filetype="
			.. filetype
			.. " "
			.. preview_buf_name
	)
	local preview_info = get_preview_info()
	if preview_info.preview_bufnr ~= nil then
		local lines = vim.split(documentation_value, "\n")
		vim.api.nvim_buf_set_lines(preview_info.preview_bufnr, 0, -1, false, lines)
		vim.bo[preview_info.preview_bufnr].filetype = filetype
		if filetype == "markdown" then
			vim.treesitter.start(preview_info.preview_bufnr)
		end
	end
end

---@param lsp_documentation string|lsp.MarkupContent
local function refresh_documentation(lsp_documentation)
	local complete_info = get_preview_info()
	--we do this check here so that we don't close the completion window
	--if we are not going to replace it.
	if is_lsp_completion_documention_empty(lsp_documentation) then
		return
	elseif complete_info.preview_bufnr == nil and get_preview_type() ~= "preview" then
		--If the documentation buffer is already closed and does not need to be
		--a "preview" window we can just create a new one without needing to schedule it
		display_documentation_popup(lsp_documentation, complete_info.selected)
	else
		--Since the preview buffer is already open we need to close it and reopen a new
		--one so that the treesitter highlights refresh. (I at least couldn't figure out a
		--better way to achieve this. I tried a lot things)
		vim.schedule(function()
			if get_preview_type() == "preview" then
				display_documentation_preview(lsp_documentation)
				return
			end

			local current_complete_info = vim.fn.complete_info({
				"selected",
				"preview_winid",
				"preview_bufnr",
			})
			if current_complete_info.preview_winid ~= nil then
				--HACK: nvim_win_hide will cause a whole new
				--buffer and window to be generated when I call nvim__complete_set
				--Since it is a new buffer, it will recalculate
				--the the treesitter highlights correctly. Without this
				--the treessiter highlights won't refresh
				vim.api.nvim_win_hide(current_complete_info.preview_winid)
			end
			display_documentation_popup(lsp_documentation, current_complete_info.selected)
		end)
	end
end

local function set_documentation(completion_item)
	if completion_item == nil then
		return
	end

	local lsp_completion_item = vim.tbl_get(completion_item, "user_data", "nvim", "lsp", "completion_item")
	local complete_info = get_preview_info()

	local preview_bufnr = complete_info.preview_bufnr
	if lsp_completion_item == nil then
		--Handle non-lsp completion items
		if preview_bufnr ~= nil and completion_item.word ~= nil and completion_item.word ~= "" then
			vim.bo[preview_bufnr].filetype = ""
		end
		return
	end

	---@type string|lsp.MarkupContent
	local lsp_documentation = lsp_completion_item.documentation
	if lsp_documentation ~= nil then
		--Handle lsp completion items with documentation already supplied
		refresh_documentation(lsp_documentation)
		return
	end

	--Handle lsp completion items that need documentation resolved
	local client_id = vim.tbl_get(completion_item, "user_data", "nvim", "lsp", "client_id")
	local client = vim.lsp.get_client_by_id(client_id)
	if client == nil then
		return
	end

	local is_resolve_support = true --TODO actually check for resolve support
	if not is_resolve_support then
		return
	end
	client.request(vim.lsp.protocol.Methods.completionItem_resolve, lsp_completion_item, function(err, result, _)
		if err ~= nil then
			vim.print(err)
			return
		end
		local documentation = result and result.documentation
		refresh_documentation(documentation)
	end)
end

local function enable_completion_resolve_documentation(bufnr)
	vim.api.nvim_create_autocmd("CompleteChanged", {
		group = vim.api.nvim_create_augroup("completion.documentation" .. bufnr, { clear = true }),
		buffer = bufnr,
		callback = function(_)
			if tonumber(vim.fn.pumvisible()) == 0 or is_documentation_disabled() then
				return
			end
			local event = vim.v.event --need to grab from vim.v.event
			set_documentation(event.completed_item)
		end,
	})
end

-------------------------------------------------------------------------------
---Setup autocmd that provides completion resolve support
vim.api.nvim_create_autocmd("LspAttach", {
	group = vim.api.nvim_create_augroup("lsp_attach", { clear = true }),
	desc = "lsp on_attach",
	callback = function(event)
		local client = vim.lsp.get_client_by_id(event.data.client_id)
		if client == nil then
			return
		end

		vim.lsp.completion.enable(true, client.id, event.buf, {
			autotrigger = true,
		})

		enable_completion_resolve_documentation(event.buf)
	end,
})

-------------------------------------------------------------------------------
---Setup dummy language server used for testing

---@alias LspServer.Handler fun(params: table?, callback: fun(err: lsp.ResponseError|nil, result: any), notify_reply_callback: fun(message_id: integer)|nil)

---@type {string: LspServer.Handler}
local handlers = {

	---@type LspServer.Handler
	[vim.lsp.protocol.Methods.initialize] = function(_, callback, _)
		local initializeResult = {
			capabilities = {
				completionProvider = {
					resolveProvider = true,
				},
			},
			serverInfo = {
				name = "dummy-lsp-server",
				version = "0.0.1",
			},
		}

		callback(nil, initializeResult)
	end,

	---@type LspServer.Handler
	[vim.lsp.protocol.Methods.textDocument_completion] = function(_, callback, _)
		---@type lsp.CompletionItem[]
		local completion_items_result = {

			{
				label = "test01",
				kind = vim.lsp.protocol.CompletionItemKind.Text,
				documentation = {
					kind = vim.lsp.protocol.MarkupKind.Markdown,
					value = "test01: with single line `` `filetype=markdown` ``. Item not resolved.",
				},
			},

			{
				label = "test02",
				kind = vim.lsp.protocol.CompletionItemKind.Text,
				documentation = {
					kind = vim.lsp.protocol.MarkupKind.Markdown,
					value = "test02: with single line `` `filetype=markdown` ``. Item not resolved. Navigating up from test03 will error, but in theory should be fixed by PR #30246",
				},
			},
			{
				label = "test03",
				kind = vim.lsp.protocol.CompletionItemKind.Text,
			},
			{
				label = "test04",
				kind = vim.lsp.protocol.CompletionItemKind.Text,
				documentation = {
					kind = vim.lsp.protocol.MarkupKind.Markdown,
					value = [[
test04: this should stay up when you navigate to test05
since test05 has no documentation in its completion item or resolve results

This item was not resolved.

some code to see treesitter highlights
\`\`\`lua
local function test()
    vim.print("Testing injected lua treesitter highlights are working")
end
\`\`\`
]],
				},
			},
			{
				label = "test05",
				kind = vim.lsp.protocol.CompletionItemKind.Text,
			},
			{
				label = "test06",
				kind = vim.lsp.protocol.CompletionItemKind.Text,
				documentation = {
					kind = vim.lsp.protocol.MarkupKind.Markdown,
					value = [[
test06: this should replace test04 documentation when navigated down from test04

This item was not resolved.

some code to see treesitter highlights
\`\`\`lua
local function test()
    vim.print("Testing injected lua treesitter highlights are working")
end
\`\`\`
]],
				},
			},
			{
				label = "test07",
				kind = vim.lsp.protocol.CompletionItemKind.Text,
				documentation = {
					kind = vim.lsp.protocol.MarkupKind.PlainText,
					value = [[
test07: this documentation is plain text so should not have treesitter highlights

This item was not resolved.

some code to see treesitter highlights are disabled
\`\`\`lua
local function test()
    vim.print("Testing that treesitter highlights are disabled since this documentation is set to be plaintext")
end
\`\`\`
]],
				},
			},
			{
				label = "test08",
				kind = vim.lsp.protocol.CompletionItemKind.Text,
				documentation = {
					kind = vim.lsp.protocol.MarkupKind.Markdown,
					value = [[
test08: switching back to markdown documentation

This item was not resolved.

some code to see treesitter highlights
\`\`\`lua
local function test()
    vim.print("Testing injected lua treesitter highlights are working")
end
\`\`\`
]],
				},
			},
			{
				label = "test09",
				kind = vim.lsp.protocol.CompletionItemKind.Text,
				--testing resolve request for plaintext documentation where documentation field is an object
			},
			{
				label = "test10",
				kind = vim.lsp.protocol.CompletionItemKind.Text,
				--testing resolve request for plaintext documentation where documentation field is a string
			},

			{
				label = "test11",
				kind = vim.lsp.protocol.CompletionItemKind.Text,
				-- documentation will be resolved
			},
			{
				label = "test12",
				kind = vim.lsp.protocol.CompletionItemKind.Text,
				-- no documentation will be resolved
			},
			{
				label = "test13",
				kind = vim.lsp.protocol.CompletionItemKind.Text,
				-- documentation will be resolved
			},
		}

		---Side Note: I schedule the callback because there seems to be a bug.
                ---If a lua-defined lsp like this is the only lsp with completion results, the
		---callback will error. The workaround is to schedule the callback.
		---I need to do more testing to verify this isn't user error like
		---not implementing the lsp spec correctly.
		---
		---I was getting the following stacktrace:
		---
		---E5108: Error executing lua Vim:E565: Not allowed to change text or change window
		-- stack traceback:
		-- [C]: in function 'complete'
		-- C:\nvim-win64\share\nvim\runtime/lua/vim/lsp/completion.lua:464: in function 'callback'
		-- C:\nvim-win64\share\nvim\runtime/lua/vim/lsp/completion.lua:389: in function 'handler'
		-- C:\nvim-win64\share\nvim\runtime/lua/vim/lsp/client.lua:681: in function 'callback'
		-- ...uments\poc\debug-neovim\issue-29225-my-init-lua\init.lua:128: in function <...uments\poc\debug-neovim\issue-29225-my-init-lua\init.lua:98>
		-- ...uments\poc\debug-neovim\issue-29225-my-init-lua\init.lua:168: in function 'request'
		-- C:\nvim-win64\share\nvim\runtime/lua/vim/lsp/client.lua:673: in function 'request'
		-- C:\nvim-win64\share\nvim\runtime/lua/vim/lsp/completion.lua:385: in function 'request'
		-- C:\nvim-win64\share\nvim\runtime/lua/vim/lsp/completion.lua:424: in function 'trigger'
		-- C:\nvim-win64\share\nvim\runtime/lua/vim/lsp/completion.lua:735: in function <C:\nvim-win64\share\nvim\runtime/lua/vim/lsp/completion.lua:725>
		vim.schedule(function()
			callback(nil, completion_items_result)
		end)
	end,

	---@type LspServer.Handler
	[vim.lsp.protocol.Methods.completionItem_resolve] = function(request, callback, _)
		if request.label == "test03" then
			local item = vim.tbl_deep_extend("force", request, {
				documentation = {
					kind = vim.lsp.protocol.MarkupKind.Markdown,
					value = [[
# test03: `` `completionItem/resolve` ``

This item was resolved.

implements [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem_resolve)
]],
				},
			})
			callback(nil, item)
		elseif request.label == "test09" then
			local item = vim.tbl_deep_extend("force", request, {
				documentation = {
					kind = vim.lsp.protocol.MarkupKind.PlainText,
					value = [[
# test09: `` `completionItem/resolve` `` with plaintext documentation kind so there should be no treesitter highlights

implements [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem_resolve)

This item was resolved.

some code to see treesitter highlights are disabled
\`\`\`lua
local function test()
    vim.print("Testing that treesitter highlights are disabled since this documentation is set to be plaintext")
end
\`\`\`
]],
				},
			})
			callback(nil, item)
		elseif request.label == "test10" then
			local item = vim.tbl_deep_extend("force", request, {
				documentation = [[
# test10: `` `completionItem/resolve` `` with plaintext documentation where documentation is a string and not an object

implements [spec](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem_resolve)

This item was resolved.

some code to see treesitter highlights are disabled
\`\`\`lua
local function test()
    vim.print("Testing that treesitter highlights are disabled since this documentation is set to be plaintext")
end
\`\`\`
]],
			})
			callback(nil, item)

		----
		elseif request.label == "test11" then
			local item = vim.tbl_deep_extend("force", request, {
				documentation = {
					kind = vim.lsp.protocol.MarkupKind.Markdown,
					value = [[
test11: this should stay up when you navigate to test12
since test12 has no documentation in its completion item or resolve results

This item was resolved.

some code to see treesitter highlights
\`\`\`lua
local function test()
    vim.print("Testing injected lua treesitter highlights are working")
end
\`\`\`
]],
				},
			})
			callback(nil, item)
		elseif request.label == "test13" then
			local item = vim.tbl_deep_extend("force", request, {
				documentation = {
					kind = vim.lsp.protocol.MarkupKind.Markdown,
					value = [[
test13: this should replace test11 documentation when navigated down from test11 and test 12

This item was resolved.

some code to see treesitter highlights
\`\`\`lua
local function test()
    vim.print("Testing injected lua treesitter highlights are working")
end
\`\`\`
]],
				},
			})
			callback(nil, item)
		end
		callback(nil, nil)
	end,
}

vim.api.nvim_create_autocmd("FileType", {
	pattern = { "lua" },
	group = vim.api.nvim_create_augroup("start_dummy_lsp", {}),
	callback = function()
		vim.lsp.start({
			name = "dummylsp",
			root_dir = vim.fn.getcwd(),
			---@type fun(dispatchers: vim.lsp.rpc.Dispatchers): vim.lsp.rpc.PublicClient
			cmd = function(_)
				---@type vim.lsp.rpc.PublicClient
				return {

					---@type fun(method: string, params: table?, callback: fun(err: lsp.ResponseError|nil, result: any), notify_reply_callback: fun(message_id: integer)|nil):boolean,integer?
					request = function(method, params, callback, notify_reply_callback)
						if handlers[method] then
							handlers[method](params, callback, notify_reply_callback)
							return true
						else
							return false
						end
					end,
					notify = function(_, _)
						return false
					end,
					is_closing = function()
						return false
					end,
					terminate = function() end,
				}
			end,
		})
	end,
})


```

## Feature Requests:

- Calling `vim.fn.complete_info({ 'preview_winid', 'preview_bufnr', })` should return back the preview_winid and preview_bufnr. Currently, complete_info only returns those values if the `selected` field is also fetched is also fetched with `vim.fn.complete_info({ 'selected', 'preview_winid', 'preview_bufnr', })`. 
  (I will create in issue for this soon)
  See https://github.com/neovim/neovim/blob/5a86360400691e55fae66d60485b61360a1d3d6c/src/nvim/insexpand.c#L2928C1-L2935C6
  Note: This may have been intentional functionality, but I don't see any documentation indicating it should behave this way. For context, these fields were implemented in [feat(complete): completeopt support popup like vim #24723](https://github.com/neovim/neovim/pull/24723).
- Modify `vim.fn.complete_info({ 'selected', 'preview_winid', 'preview_bufnr', })` to return back the winid and bufnr of the documentation window regardless of if it is the floating documentation window or preview documentation window
  (I will create an issue for this soon)
  Note: the help documentation clearly says that preview_winid and preview_bufnr will be the of the "floating preview window" so changing this may be a minor breaking. The one other concern with implementing this is how to handle the situation where the preview window's contents no longer contains the documentation. I think initial implementation should err on simplicity and just return the bufnr and winid without any fancy validation. For context, these fields were originally implemented in [feat(complete): completeopt support popup like vim #24723](https://github.com/neovim/neovim/pull/24723)
- Add an `infotype` field to completion item that would contain the filetype that the documentation buffer should be set to. Imo, this would be really nice because of the awkward desire path I described in the pain points section. My init.lua implementation doesn't use the completion info field other than to check it is empty. Since I already have to dig into the lsp results to get the filetype, it is just more convenent to grab the documentation contents from the same place. Side-Note: After the tresitter bugs are fixed, I might refactor my init.lua to see if that desire path is still there since I can imagine that making a difference.

## Bugs:

- [treesitter: Invalid 'end_row': out of range on CompleteChanged #30610](https://github.com/neovim/neovim/issues/30610) This is the one bug that I was not able to work around in my init.lua implementation but this will be fixed as soon as glepnir's PR [fix(popup): wrong extmark data sync when lines changed in popup preview #30246](https://github.com/neovim/neovim/pull/30246) is merged.
- Treesitter highlighter would not redraw highlights in the documentation popup window when the contents changed. The only work around I could find was closing popup window with `vim.api.nvim_win_hide` and recreating it with `vim.api.nvim__complete_set` I suspect that glepnir's PR [fix(popup): wrong extmark data sync when lines changed in popup preview #30246](https://github.com/neovim/neovim/pull/30246) may also fix this so I will test that before creating a separate issue.
- My init.lua would throw an error if I did not schedule the completion result callback when no secondary lsp was attached. I need do some more testing and make sure this is not an user error before I create an issue for it.

## `nvim__complete_set` Bugs and Features:

I found some bugs with `vim.api.nvim__complete_set`, but since it was demoted to an experimental API in [fix(api): mark nvim\_\_complete_set as experimental #28579](https://github.com/neovim/neovim/pull/28579) I would like feedback as to whether or not I should create issues for these. I saw glepnir [commented](https://github.com/neovim/neovim/pull/28579#issuecomment-2085164965) that `nvim__complete_set` won't be needed after [Completion API: completion handlers, Lua callbacks on completion events #16488](https://github.com/neovim/neovim/issues/16488), but I couldn't figure out which part would be the replacement for `nvim__complete_set`. Since that issue is still open and seems to be a larger design initiative, my preference would be to do small bug fixes and improvements to `vim.api.nvim__complete_set` for the time being. Then refactor whenever a replacement is implemented.

- Bug: `nvim__complete_set` does not respect `vim.o.completeopt` and will always use the popup window even when completopt has the preview option. Fixing this would significantly simplify my `completionItem/resolve` implementation.
- Bug: `nvim__complete_set` will crash after several calls when completeopt has preview option.
- Feature: Add an infotype option to `nvim__complete_set` options (imo, `vim.bo[preview_bufnr].filetype` would probably be sufficient. However, if we still have issues with treesitter highlights, we may want this field so that we can reuse the fixes from [fix(popup): wrong extmark data sync when lines changed in popup preview #30246](https://github.com/neovim/neovim/pull/30246))

## Possible Alternative Features:

There are other features that would simplify my `completionItem/resolve` implementation that I considered as possibilities but decided that I prefer my other suggestions. I won't be creating issues for these unless someone with more insight on the vision thinks I should.

- Add a better way to lazily determine the completion item's info field. This could be an alternative to using `nvim__complete_set`. Possibilities include:
  - vim.bo.omnidocfunc
  - Allow info field to be a lua function
  - Add a CompleteChangedPre event that gives the user a chance to modify completed_item before the documentation is displayed (may not be that useful as it sounds since running the lsp request asynchronously would defeat the purpose of this autocmd)

## Conclusion

The C code for most of the changes I proposed (excluding the "Possible Alternative Features" section) would not be difficult to implement. The existing code is fairly easy to read and has good tests. If they are in line with the vision, I feel comfortable creating PRs for these.

TLDR: I created a possible lua only implementation for LSP `completionItem/resolve` support for completion documentation. However, it could be simplified with changes to the C code. I would like feedback on if this approach and my suggestions are in line with the vision.

Lastly, thanks for all the great work on the LSP completion. It's been exciting playing around with it.

**Comment by glepnir** [comment]
I suggest you try #30246 before comment... It fixes most of the crashes related to ts. Secondly, nvim__complete_info only works for popup. preview is not supported and will not be supported. As far as I know, vim also does not support you setting info in a callback and then creating a normal preview window

**Comment by justinmk** [comment]
> I found some bugs with `vim.api.nvim__complete_set`, but since it was demoted to an experimental API in [fix(api): mark nvim__complete_set as experimental #28579](https://github.com/neovim/neovim/pull/28579) I would like feedback as to whether or not I should create issues for these

@crwebb85 yes, feedback on experimental APIs is welcome, and actually a crucial mechanism of maturing experimental APIs (we need feedback on them, how they are used, etc.)

**Comment by justinmk** [comment]
Initial impl is in https://github.com/neovim/neovim/pull/32820 but @crwebb85 it looks like you have some ideas about improvements that would be welcome.

Somewhat related, we could really use a "debounce" util to avoid reimplementing that over and over.
